Mathlib/Topology/Instances/EReal.lean

```diff
@@ -171,6 +170,154 @@ theorem tendsto_nhds_bot_iff_real {α : Type*} {m : α → EReal} {f : Filter α
+  rw [← @limsup_const EReal α _ f _ (a + b)]
+  apply limsup_le_limsup (Eventually.mp (Eventually.and (eventually_lt_of_limsup_lt ha)
+    (eventually_lt_of_limsup_lt hb)) (eventually_of_forall _))
+  simp only [Pi.add_apply, and_imp]
+  intro x
+  exact fun ux_lt_a vx_lt_b ↦ add_le_add (le_of_lt ux_lt_a) (le_of_lt vx_lt_b)
+
+theorem limsup_add_bot_of_ne_top {α : Type _} {f : Filter α} {u : α → EReal} {v : α → EReal}
+    (h : limsup u f = ⊥) (h' : limsup v f ≠ ⊤) : limsup (u + v) f = ⊥ := by
+  apply le_bot_iff.1
+  apply (le_iff_le_forall_real_gt ⊥ (limsup (u + v) f)).1
+  intro x
+  rcases exists_between_coe_real (h'.lt_top) with ⟨y, ⟨hy, _⟩⟩
+  rw [← sub_add_cancel x y, coe_add (x - y) y, coe_sub x y]
+  intro _
+  apply @limsup_add_le_lt₂ α f u v (x - y) y _ hy
+  rw [h, ← coe_sub x y]
+  exact bot_lt_coe (x - y)
+
+theorem limsup_add_le_add_limsup {α : Type _} {f : Filter α} {u v : α → EReal}
+    (h : limsup u f ≠ ⊥ ∨ limsup v f ≠ ⊤) (h' : limsup u f ≠ ⊤ ∨ limsup v f ≠ ⊥) :
+    limsup (u + v) f ≤ (limsup u f) + (limsup v f) := by
+  rcases eq_bot_or_bot_lt (limsup u f) with (u_bot | u_nbot)
+  · rcases h with (u_nbot | v_ntop)
+    · exfalso; exact u_nbot u_bot
+    · rw [limsup_add_bot_of_ne_top u_bot v_ntop]; exact bot_le
+  rcases eq_bot_or_bot_lt (limsup v f) with (v_bot | v_nbot)
+  · rcases h' with (u_ntop | v_nbot)
+    · rw [add_comm, limsup_add_bot_of_ne_top v_bot u_ntop]; exact bot_le
+    · exfalso; exact v_nbot v_bot
+  rcases eq_top_or_lt_top (limsup v f) with (v_top | v_ntop)
+  · rw [v_top, add_top_of_ne_bot (ne_of_gt u_nbot)]; exact le_top
+  have limsup_v_real := coe_toReal (ne_of_lt v_ntop) (ne_of_gt v_nbot)
+  apply (le_iff_le_forall_real_gt _ _).1
+  intros x hx
+  rcases lt_iff_exists_real_btwn.1 hx with ⟨y, ⟨sum_lt_y, y_lt_x⟩⟩
+  have key₁ : limsup u f < (y - limsup v f) := by
+    apply lt_of_eq_of_lt _ (sub_lt_sub_of_lt_of_le sum_lt_y (le_of_eq (Eq.refl (limsup v f)))
+      (ne_of_gt v_nbot) (ne_of_lt v_ntop))
+    rw [← limsup_v_real, add_sub_cancel_right]
+  have key₂ : limsup v f < limsup v f + x - y := by
+    rw [← limsup_v_real]; norm_cast; norm_cast at y_lt_x; linarith
+  apply le_of_le_of_eq (limsup_add_le_lt₂ key₁ key₂)
+  rw [← limsup_v_real]; norm_cast; linarith
+
+lemma liminf_add_ge_gt₂ {α : Type _} {f : Filter α} {u v : α → EReal} {a b : EReal}
+    (ha : a < liminf u f) (hb : b < liminf v f) : a + b ≤ liminf (u + v) f := by
+  rcases eq_or_neBot f with (rfl | _); simp only [liminf_bot, le_top]
+  rw [← @liminf_const EReal α _ f _ (a + b)]
+  apply liminf_le_liminf (Eventually.mp (Eventually.and (eventually_lt_of_lt_liminf ha)
+    (eventually_lt_of_lt_liminf hb)) (eventually_of_forall _))
+  simp only [Pi.add_apply, and_imp]
+  intro x
+  exact fun ux_lt_a vx_lt_b ↦ add_le_add (le_of_lt ux_lt_a) (le_of_lt vx_lt_b)
+
+lemma liminf_add_top_of_ne_bot {α : Type _} {f : Filter α} {u : α → EReal} {v : α → EReal}
+    (h : liminf u f = ⊤) (h' : liminf v f ≠ ⊥) : liminf (u + v) f = ⊤ := by
```

Comment:
Again, just dualize with [EReal.negOrderIso](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Real/EReal.html#EReal.negOrderIso)

Reply:
I think I can do it but probably the proof wouldn't be shorter than the current one.